Register DisplayDrivers with ContentOptions

deanmarcussen opened the pull request:
Far too late to be wanting to get this in now, but further to our conversation @sebastienros here's some ideas for registering DisplayDriver Types in the new ContentOptions, with a factory to resolve them from DI as required, rather than resolving all of them at once and iterating.

Just ideas for the future, but if you like the concept, then this would work for Handlers, as well.

Did think about using a dictionary of Func<> but the factory seemed better

Also will probably need to be changed to a Lookup to support multiple driver registrations against the same part, eg CommonPart Drivers

jtkech commented on src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,27 @@
+            contentPartBuilder.Services.AddScoped<TContentPartDisplayDriver>();
+            contentPartBuilder.Services.Configure<ContentOptions>(o =>
+                o.WithContentPartFactoryType("displaydriver", contentPartBuilder.ContentPartType, displayDriverType));
+            return contentPartBuilder;
```
Just did a quick look so not sure, will see more in details asap

Maybe use nameof `IContentPartDisplayDriver`, or use `Type` keys and here typeof `IContentPartDisplayDriver`.

jtkech commented on src/OrchardCore/OrchardCore.ContentManagement.Abstractions/ContentOptions.cs:
```diff
@@ -29,6 +30,12 @@ public ContentPartOption AddContentPart(Type contentPartType)
+        public void WithContentPartFactoryType(string factoryName, Type contentPartType, Type factoryType)
+        {
+            var option = _contentParts.FirstOrDefault(x => x.Type.Name == contentPartType.Name);
+            option.WithFactoryType(factoryName, factoryType);
```
Just did a quick look so not sure, will see more in details asap.

I think here you can just do `x.Type == contentPartType`.

jtkech commented on src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,27 @@
+            var displayDriverType = typeof(TContentPartDisplayDriver);
+
+            if (displayDriverType.BaseType != null && !Array.Exists(displayDriverType.BaseType.GenericTypeArguments, x => x == contentPartBuilder.ContentPartType))
+            {
```
Here just as i remember.

Maybe just `BaseType.GenericTypeArguments[0]`. Hmm, yes but if empty it would fail without your message.

So, what you do if `BaseType` is null? Do we need to check this? Is it possible to tied a part to a non generic driver?

Hmm, maybe (but it depends on the above)

    if (!BaseType.IsGenericType || BaseType.GenericTypeArguments[0] != ContentPartType)

With a null check on the `BaseType` would be `!(BaseType?.IsGenericType ?? false)`

jtkech commented on src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,27 @@
+            if (displayDriverType.BaseType != null && !Array.Exists(displayDriverType.BaseType.GenericTypeArguments, x => x == contentPartBuilder.ContentPartType))
+            {
+                throw new ArgumentException("The display driver type must inherit from " + contentPartBuilder.ContentPartType.Name);
+            }
```
The message would specify to inherit from `ContentPartDisplayDriver<{ContentPartType}>`

deanmarcussen commented:
@jtkech Thanks for taking a look last night.

I had some time this morning so I refactored, based on your suggestions, and some other things that I wanted to try.

- Removed the lookup dictionary on `ContentOptions`, and implemented `ContentDisplayOptions` as a complement too it. Felt it was a bit too leaky, and as this is all about perf, better to have an actual property to use.
- Renamed factories to Resolvers. Much better convention
- Removed weird generic type checking. Better to just rely on `where T : IContentPartDisplayDriver`
- Made it support multiple drivers and handlers for the same part, including registering seperately, if for example, they are part of a feature.
- Implemented for all Fields, and Handlers as well, to see if it benchmarks any better.

I guess this PR is all about performance, so if @sebastienros feels like benchmarking it, that would be the only way to see if what I've done is even remotely useful?

carlwoodhouse commented:
if nothing else its much more intuitive :)

deanmarcussen commented:
@sebastienros Any feedback on this idea?

deanmarcussen commented:
@jtkech I went through this again to add some logging, and am happy with how it looks, if you had any thoughts they'd be appreciated

hishamco commented on src/OrchardCore/OrchardCore.ContentManagement/Handlers/ContentPartHandlerCoordinator.cs:
```diff
@@ -328,7 +328,10 @@ public override async Task PublishingAsync(PublishContentContext context)
                     {
-                        await _partHandlers.InvokeAsync(handler => handler.PublishingAsync(context, part), Logger);
+                        await _partHandlers.InvokeAsync(async handler => {
+                            Logger.LogWarning("The handler '{Handler}' should not be registerd as IContentPartHandler. Use WithHandler<T> instead.", handler.GetType());
```
Localize the log message

hishamco commented on src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentItemDisplayCoordinator.cs:
```diff
@@ -333,6 +333,7 @@ public async Task BuildEditorAsync(ContentItem contentItem, BuildEditorContext c
                     {
                         await _fieldDisplayDrivers.InvokeAsync(async contentDisplay =>
                         {
+                            Logger.LogWarning("The display driver '{DisplayDriver}' should not be registerd as IContentFieldDisplayDriver. Use WithDisplayDriver<T> instead.", contentDisplay.GetType());
```
Localize the log message

hishamco commented on src/OrchardCore/OrchardCore.ContentManagement.Abstractions/ContentOptions.cs:
```diff
@@ -30,33 +31,35 @@ internal ContentPartOption AddContentPart(Type contentPartType)
         }
 
-        internal ContentFieldOption AddContentField<T>() where T : ContentField
+        public ContentFieldOption AddContentField<T>() where T : ContentField
```
```csharp
public ContentFieldOption AddContentField<T>() where T : ContentField
    => AddContentField(typeof(T));
```

jtkech commented:
@deanmarcussen okay no problem, i will look at it asap, but so many things ;)

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentDisplayOptions.cs:
```diff
@@ -0,0 +1,62 @@
+{
+    public class ContentDisplayOptions
+    {
+
```
@hishamco Thanks for the review. 
But you reviewed the code from 3 weeks ago. 
I'll resolve these for now

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentItemDisplayCoordinator.cs:
```diff
@@ -77,22 +84,46 @@ public async Task BuildDisplayAsync(ContentItem contentItem, BuildDisplayContext
+                        foreach (var displayDriver in _partDisplayDrivers)
                         {
-                            InvokeExtensions.HandleException(ex, Logger, displayDriver.GetType().Name, "BuildDisplayAsync");
+                            try
```
@sebastienros on your recent pr https://github.com/OrchardCMS/OrchardCore/pull/4887 you've reduced the allocation pattern for using the invoke extensions.
Just noting that the current `BuildDisplayAsync` method doesn't use the 
`await partDisplayDrivers.InvokeAsync(async (driver, part, typePartDefinition, context) =>` technique at all, it is only the `BuildEditorAsync` that used it previously.
So if there is an opportunity to reduce allocations on `
BuildDisplayAsync` (regardless of whether this PR goes through), perhaps we should use them?

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentItemDisplayCoordinator.cs:
```diff
@@ -139,19 +170,44 @@ public async Task BuildDisplayAsync(ContentItem contentItem, BuildDisplayContext
                     {
-                        foreach (var displayDriver in _fieldDisplayDrivers)
+                        var fieldDisplayDrivers = _contentFieldDisplayDriverResolver.GetDisplayDrivers(contentPartFieldDefinition.FieldDefinition.Name);
+                        // For backward compatability check for Any().
```
This `_contentFieldDisplayDriverResolver.GetDisplayDrivers(contentPartFieldDefinition.FieldDefinition.Name);`
would be updated to something like
`_contentFieldDisplayDriverResolver.GetDisplayDrivers(contentPartFieldDefinition.FieldDefinition.Name, displayMode);`

or editor, depending on the place

deanmarcussen commented:
ok @sebastienros I reimplemented this with explicit registrations for display modes and editors.

By default registration is `*` to indicate all display modes.

Other options include registering as 
- `//my-display-mode` (standard display mode and my-display-mode) as standard = `String.Empty` (could mutate that to read better i.e. /standard/my-display-mode`)
- `my-display-mode`
- `*` all display modes
- `my-display-mode/my-other-display-mode`

I have left them as mutable `HashSets` so registration can be mutated by a user, being able to alter the default registrations if they are implementing a custom editor and want to change the default registrations.

**Note** parts only have custom editors , not display modes.

deanmarcussen commented:
Updated this after our conversation Seb 

Here are some examples of how registration could be done for fields

```
// Taxonomy Field - Standard registration, will automatically add a Func return true to editor and displaymode
services.AddContentField<TaxonomyField>()
  .UseDisplayDriver<TaxonomyFieldDisplayDriver>();

// Alter display mode in third party module. Run normal driver for standard mode. Run Tag driver for Tags
services.AddContentField<TaxonomyField>() //can be called multiple times to access builder again
  .ForDisplayMode<TaxonomyFieldDisplayDriver>(d => String.IsNullOrEmpty(d) ? true : false)
  .ForDisplayMode<TagFieldDisplayDriver>(d => String.Equals(d, "Tags", StringComparison.OrdinalIgnoreCase) ? true : false);

// Alter editor in third party module
services.AddContentField<TaxonomyField>()
  .ForEditor<TaxonomyFieldDisplayDriver>(e => String.IsNullOrEmpty(e) ? true : false)
  .ForEditor<TagFieldDisplayDriver>(e => String.Equals(e, "Tags", StringComparison.OrdinalIgnoreCase) ? true : false);

// Alter both display mode and editor in third party module
services.AddContentField<TaxonomyField>()
  .UseDisplayDriver<TaxonomyFieldDisplayDriver>(f => String.IsNullOrEmpty(f) ? true : false)
  .UseDisplayDriver<TagFieldDisplayDriver>(d => String.Equals(d, "Tags", StringComparison.OrdinalIgnoreCase) ? true : false);

// Totally remove the default driver (saves running a func against it), and use a custom driver.
services.AddContentField<TaxonomyField>()
  .RemoveDisplayDriver<TaxonomyFieldDisplayDriver>()
  .UseDisplayDriver<MyCustomFieldDisplayDriver>();
```

Handlers I moved to `.Add` and `.Remove`
```
// Taxonomy Part
  services.AddContentPart<TaxonomyPart>()
    .UseDisplayDriver<TaxonomyPartDisplayDriver>()
    .AddHandler<TaxonomyHandler>();

// Remove handler in third party module and replace with a new handler
services.AddContentPart<TaxonomyPart>()
  .UseDisplayDriver<TaxonomyPartDisplayDriver>()
  .RemoveHandler<TaxonomyHandler>()
  .AddHandler<CustomTaxonomyHandler>();  
```

So let me know if you agree with the design and like the naming conventions (easy to change)
